src/Microsoft.ML.TimeSeries/SequentialTransformerBase.cs

```diff
@@ -0,0 +1,299 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Data.IO;
+using Microsoft.ML.Runtime.Internal.Utilities;
+using Microsoft.ML.Runtime.Model;
+using Microsoft.ML.Runtime.Api;
+using Microsoft.ML.Core.Data;
+
+namespace Microsoft.ML.Runtime.TimeSeriesProcessing
+{
+    /// <summary>
+    /// The base class for sequential processing transforms. This class implements the basic sliding window buffering. The derived classes need to specify the transform logic,
+    /// the initialization logic and the learning logic via implementing the abstract methods TransformCore(), InitializeStateCore() and LearnStateFromDataCore(), respectively
+    /// </summary>
+    /// <typeparam name="TInput">The input type of the sequential processing.</typeparam>
+    /// <typeparam name="TOutput">The dst type of the sequential processing.</typeparam>
+    /// <typeparam name="TState">The state type of the sequential processing. Must be a class inherited from StateBase </typeparam>
+    public abstract class SequentialTransformerBase<TInput, TOutput, TState> : ITransformer, ICanSaveModel
+       where TState : SequentialTransformerBase<TInput, TOutput, TState>.StateBase, new()
```

Comment:
>TState [](start = 13, length = 6)

I tried to understand why we even need `TState` as a type parameter any more

Reply:
The state as type is supplied not only from SSA & IID but also from numerous other timeseries transforms like PValue, SlidingWindow etc.
It would be better to look into how to get rid of it in separate PR where we plan to move the rest of timeseries transforms to estimators.
Just out of  curiosity I tried to remove the TState in this PR, the problem arose in how to pass state type into the Mapper object as we now dont have it passed as template param. I prefer to solve this in separate PR


---
In reply to: [226790319](https://github.com/dotnet/machinelearning/pull/1254#discussion_r226790319) [](ancestors = 226790319)